### PR TITLE
[Caffe2] Update Im2ColNd functions to make all params on host

### PR DESCRIPTION
**Pull request by xiaomengy**
Some benchmark:

Conv3D on N = 128, C = 128, T = 32, H = 32, W = 32, M = 128, stride = 1, kernel = 3, dilation = 1, pad = 1
Foward: 7.4s -> 1.67s
Backward: 10+s -> 3.8s

### CODE REVIEW

**Review by yinghai — COMMENTED** [review]
What's the key idea of the performance improvement?

**Comment by BIT-silence** [comment]
The main purpose of this PR is make the math functions like Im2ColNd and Col2ImNd with all the parameters on host instead of on device. Then there will be no need for the ops to upload the host to extra tensors while calling such functions. 
At the same time, we found the origin implementation of Im2ColNd and Col2ImNd on GPU are not very well. The main problem is the size of col tensor is kernel_size * output_image_size. While our current implementation only run multi-thread on kernel_size dim, all the computation on output_image_size dim are in a single thread. That's the main performance issue which also be fixed in this PR>

File: caffe2/utils/math_gpu.cu [hunk 1]
```diff
@@ -1579,436 +1650,413 @@ __global__ void col2im_gpu_kernel_nhwc(const int n, const T* data_col,
...
+      d_offset[d_i] = offset_i % kernel_shape.data[d_i];
+      offset_i /= kernel_shape.data[d_i];
     }
-    // Initialize offset, computed in the loop below, with intermediate
-    // computations used to compute the spatial indices.
-    int offset = index;
-    for (i = num_axes - 1; i >= 0; --i) {
-      if (i < num_axes - 1) {
-        offset /= shared_kernel_shape[i + 1];
+    for (int j = threadIdx.x; j < inner_size; j += blockDim.x) {
```
Comment (yinghai) [inline caffe2/utils/math_gpu.cu]: Ok. I can see the double parallel loop here.

**Review by yinghai — APPROVED** [review]
Looks good to me. One nit is that this turns out to be a big diff. While the key change isn't big, much of the diffs are about formatting and cleaning up. I would suggest split them to keep the idea simple. Or have more comments in the PR to highlight the idea.

**Pull request merged**